File: Python/swapview.py
#!/usr/bin/env python3

from __future__ import print_function
from __future__ import division

import os
import re

format = "%7s %9s %s"
totalFmt = "Total: %10s"

def filesize(size):
  units = 'KMGT'
  left = abs(size)
  unit = -1
  while left > 1100 and unit < 3:
    left /= 1024
    unit += 1
  if unit == -1:
    return '%dB' % size
  else:
    if size < 0:
      left = -left
    return '%.1f%siB' % (left, units[unit])

def getSwapFor(pid):
  try:
    comm = open('/proc/%s/cmdline' % pid).read()
    if comm and comm[-1] == '\x00':
      comm = comm[:-1]
    comm = comm.replace('\x00', ' ')
    with open('/proc/%s/smaps' % pid) as f:
      s = sum(int(m.group(1)) for m in
              re.finditer(r'\nSwap:\s+(\d+)', f.read()))
    return pid, s * 1024, comm
  except (IOError, OSError):
    return pid, 0, ''

def getSwap():
  ret = []
  for pid in os.listdir('/proc'):
    if pid.isdigit():
      s = getSwapFor(pid)
      if s[1] > 0:
        ret.append(s)
  ret.sort(key=lambda x: x[1])
  return ret

def main():
  results = getSwap()
  print(format % ('PID', 'SWAP', 'COMMAND'))
  for pid, swap, comm in results:
    print(format % (pid, filesize(swap), comm))
  t = filesize(sum(x[1] for x in results))
  print(totalFmt % t)

if __name__ == '__main__':
  main()

File: NodeJS/swapview.js
#!/usr/bin/env node
"use strict"
const sprintf = require('sprintf-js').sprintf
const fs = require("fs")

function filesize(size) {
  const units = "KMGT"
  let left = size
  let unit = -1
  for (; left > 1100 && unit < 3; unit++) {
    left /= 1024
  }
  if (unit === -1) {
    return sprintf("%dB", size)
  } else {
    if (size < 0) left = -left
    return sprintf('%.1f%siB', left, units[unit])
  }
}

function getSwapFor(pid){
  try {
    let comm = fs.readFileSync(`/proc/${pid}/cmdline`, "utf-8")
    if(comm[comm.length-1] === '\0'){
      comm = comm.slice(0, -1)
    }
    comm = comm.replace(/\0/g," ")
    let s = 0
    const smaps = fs.readFileSync(`/proc/${pid}/smaps`, "utf-8")
    const re = /\nSwap:\s+(\d+)/g
    let result
    while (result = re.exec(smaps)) {
      s += result[1] | 0
    }
    return [pid, s*1024, comm]
  } catch(e) {
    if(e.code === 'EACCES' || e.code === 'ENOENT') {
      return [pid, 0, ""]
    } else {
      throw e
    }
  }
}

function getSwap() {
  const ret = []
  fs.readdirSync("/proc").forEach(spid => {
    const pid = spid | 0
    if (pid > 0) {
      const s = getSwapFor(pid)
      if (s[1] > 0) {
        ret.push(s)
      }
    }
  })
  ret.sort((a, b) => a[1] - b[1])
  return ret
}

const results = getSwap()
console.log(sprintf("%7s %9s %s", "PID", "SWAP", "COMMAND"))
let t = 0
results.forEach(s => {
  console.log(sprintf("%7s %9s %s", s[0], filesize(s[1]), s[2]))
  t += s[1]
})

console.log(sprintf("Total: %8s", filesize(t)))
